project/MultiScalaProject.scala

```diff
@@ -224,8 +220,56 @@ object MultiScalaProject {
       name,
       projects,
       dependsOnSourceInIDE = additionalIDEScalaVersions.nonEmpty
-    ).settings(
-      sourceDirectory := baseDirectory.value.getParentFile / "src"
     )
   }
+
+  private def srcDir(root: File, base: File) = root / base.getPath / "src"
+
+  private def sharedSourceDirsForConfig(
+      bases: Seq[File],
+      subdir: String,
+      conf: Configuration
+  ) = {
+    conf / unmanagedSourceDirectories ++= {
+      val dirs =
+        bases.map(x => srcDir((ThisBuild / baseDirectory).value, x) / subdir)
+      val vers = CrossVersion.partialVersion(scalaVersion.value) match {
+        case Some((2, 12)) => Seq("2", "2.12")
+        case Some((2, 13)) => Seq("2", "2.13", "2.13+")
+        case Some((3, _))  => Seq("3", "2.13+")
+        case _ => sys.error(s"Unsupported Scala version: ${scalaVersion}")
+      }
+      ("scala" +: vers.map("scala-" + _)).flatMap(v => dirs.map(_ / v))
+    }
+  }
+
+  private def sharedSourceDirs(initialBases: Seq[File]) = {
+    val bases = initialBases ++ getPlatformSourceDirs(initialBases)
+    Def.settings(
+      sharedSourceDirsForConfig(bases, "main", Compile),
+      sharedSourceDirsForConfig(bases, "test", Test)
+    )
+  }
+
+  private def getPlatformSourceDirs(bases: Seq[File]): Seq[File] = {
+    val builder = Seq.newBuilder[File]
+
+    def platformDir(platform: String): Unit =
+      bases.foreach(builder += _ / platform)
+
+    def platformDirs(platforms: String*): Unit =
+      platforms.foreach(platformDir)
+
+    val osName = sys.props("os.name").toLowerCase
+    if (osName.startsWith("win")) platformDir("windows")
+    else {
+      platformDir("unix")
+      if (osName.contains("linux")) platformDir("linux")
+      else if (osName.contains("mac")) platformDirs("macos", "bsd")
+      else if (osName.contains("freebsd")) platformDirs("freebsd", "bsd")
+    }
+
+    builder.result()
+  }
```

Comment:
This thing should not be included - we cannot have platform specific sources, because we don't cross-compile for different OS/CPU, don't publish platform specific sources or have compilation from sources by the users - all publishing is done on Ubuntu/x86_64. 
Every code that is platform specific needs to be guarded with linktime-condition so that we can erase untaken pathes when linking based on the NativeConfig